File: 윤상우/Week_11/P72411_메뉴리뉴얼.java
package temp;
import java.util.*;
public class P72411_메뉴리뉴얼 {
    static int [] selection;
    static HashMap<String, Integer> map;
    static List<String> answer;

    static void combination(String s, int start, int r, int R){
        if(r==R){
            String str = "";
            for(int i=0; i<R; i++){
                str += s.charAt(selection[i]);
            }
            char[] StringtoChar = str.toCharArray();
            Arrays.sort(StringtoChar);
            str = new String(StringtoChar);

            if(map.containsKey(str)){
                int num = map.get(str);
                map.remove(str);
                map.put(str, num+1);
            }else{
                map.put(str,1);
            }
            return;
        }

        for(int i=start; i<s.length(); i++){
            selection[r] = i;
            combination(s,i+1,r+1,R);
        }
    }

    public List<String> solution(String[] orders, int[] course) {
        answer = new ArrayList<>();

        for(int i : course){
            selection = new int [i];
            map = new HashMap<String, Integer>();
            int max = 0;

            for(String s : orders){
                combination(s,0,0,i);
            }

            Set<String> keySet = map.keySet();
            for(String key : keySet){
                max = Math.max(max, map.get(key));
            }

            if(max >=2){
                for(String key : keySet){
                    if(map.get(key) == max){
                        answer.add(key);
                    }
                }
            }
        }
        Collections.sort(answer);
        return answer;
    }
}

File: 곽승규/Week_11/P72411_메뉴리뉴얼.py
from itertools import combinations
from collections import Counter


def solution(orders, course):
    answer = []

    for c in course:
        temp = []
        for order in orders:
            combination = combinations(sorted(order), c)  # 정렬해서 조합 생성
            combination_list = list(combination)  # 조합을 리스트로 변환
            temp += combination_list  # 하나의 리스트에 다 저장

        counter = Counter(temp)  # Counter(temp)를 하면 위의 조합별로 횟수를 계산

        if len(counter) != 0 and max(counter.values()) != 1:
            for i in counter:
                if counter[i] == max(counter.values()):
                    answer += ["".join(i)]
    return sorted(answer)
